Added RATS (Request for Answer To Select)

JPG-Consulting opened the pull request:
Added RATS for ISO-14443-4 PICC

Rotzbua commented:
Thank, looks good.
@JPG-Consulting Please can you give me a short example use case? Sorry it has been a long time since I worked with rfid.

JPG-Consulting commented:
@Rotzbua For example in the dump info example:
https://github.com/miguelbalboa/rfid/blob/master/examples/DumpInfo/DumpInfo.ino

You can call it just before the dump (To avoid the HaltA being called in the dump method).Just check the card sak id 0x20 or the card type is PICC_TYPE_ISO_14443_4.

ISO 1443-4 returns a 16 byte ATS therefore something like:

``
byte ats[16];
byte atsLength;

mfrc522.PICC_RATS(ats, &atsLength);

should do it. You cound check the returned status code is STATUS_OK and then print out the ATS.

Works fine with a MIFARE DESFire... but I'm still having troubles sending the 0x60 (GET VERSION) byte to the card :( Right now I've fully stripped the library to the minimal to try and figure out what's wrong with that step... and probably the code can be improved a little by storing the ATS into a structure as the bytes are defined on a PDF.

JPG-Consulting commented:
Ok. I've worked a little on this commit.

I've automated the RATS on `ReadCardSerial`.

I've also added a new structure called `CardInfo` to hold the `Uid` and some more information such as the ATQA. Now a MIFARE DESFire card can be identified although nothing more is done with this card except for ISO-14443-4 RATS. Only cards with the sixth bit of the SAK turned on (ISO-14443-4 cards) will request the ATS.

All the older functions have been kept for backward compatibility.

The only modification to the dump code is passing `mfrc522.card` instead of `mfrc522.uid`. The code is shown below:

    #include <SPI.h>
    #include <MFRC522.h>
    
    #define SS_PIN 10
    #define RST_PIN 9
    MFRC522 mfrc522(SS_PIN, RST_PIN);  // Create MFRC522 instance.

    void setup() {
      Serial.begin(9600); // Initialize serial communications with the PC
      SPI.begin();      // Init SPI bus
      mfrc522.PCD_Init(); // Init MFRC522 card
      Serial.println("Scan PICC to see UID and type...");
    }

    void loop() {
      // Look for new cards
      if ( ! mfrc522.PICC_IsNewCardPresent()) {
        return;
      }
      
      Serial.println("Card is present....");
      
      // Select one of the cards
      if ( ! mfrc522.PICC_ReadCardSerial()) {
        return;
      }
    
      // Dump debug info about the card. PICC_HaltA() is automatically called.
      mfrc522.PICC_DumpToSerial(&(mfrc522.card));
    }

Rotzbua commented:
@JPG-Consulting It would be nice if you extend `keywords.txt` with your new functions. Thanks.

JPG-Consulting commented:
@Rotzbua done.

However I'm not sure it is 100% correct as I'm not sure how `keyword.txt` works or even the use of it.

Also, some functions have been almost copy-pasted. For example:

    static PICC_Type PICC_GetType(byte sak);
    static PICC_Type PICC_GetType(CardInfo *card);
    
    void PICC_DumpToSerial(CardInfo *card);
    void PICC_DumpToSerial(Uid *uid);
    
    void PICC_DumpDetailsToSerial(CardInfo *card);
    void PICC_DumpDetailsToSerial(Uid *uid);

I left the old methods to keep backward compatibility and not having to rewritte all of the samples using the `CardInfo` structure. As mentioned `CardInfo` is simplly a structure to hold further information that the one in `Uid` such as the `ATQA` in order to better identify the car type (Required if we want to identify a MIFARE DESFire or any other specific ISO-14443 card).

Rotzbua commented on MFRC522.cpp:
```diff
@@ -1870,6 +2077,34 @@ bool MFRC522::PICC_IsNewCardPresent() {
+			// Command is oriented to this type of card
+			PICC_Type piccType = PICC_GetType(&card);
+			if (piccType == PICC_TYPE_MIFARE_DESFIRE) {
+				result = PICC_PPS(0x00, 0x11, 0x00);
```
Why result? Not used after

JPG-Consulting replied:
That's a good question! I don't remember. 

I guess i was going to do the enhancements I mention on another comment just below... I think I really jumped a little just to get it working up to the DESFire code and the result was just left behind. I think that's the reason because:

    // I'm only taking care of MIFARE DESFire so the PPS
    // Command is oriented to this type of card
    PICC_Type piccType = PICC_GetType(&card);
    if (piccType == PICC_TYPE_MIFARE_DESFIRE) {

Really if the PPS is present it shoud be sent for ISO/IEC-14443 cards and not only DESFire. I guess I turned a bit lazy and instead of adding those check I went into a known condition where it is a DESFire. The PPS has also been hardcoded and therefore it is telling the card to use the same speed as the MFRC522 has setup by default. Therefore if it accepts the command there is no need to tell the MFRC522 to change it's speed and if it fails probablly PPS was not supported (That check should come from the ATS answer). i guess i just got lazy and left behind some clues of my lazyness. I'll look into it :)

Rotzbua commented on MFRC522.cpp:
```diff
@@ -1250,6 +1306,36 @@ const __FlashStringHelper *MFRC522::GetStatusCodeName(MFRC522::StatusCode code	/
+	// 3.2 Coding of Select Acknowledge (SAK)
+	// ignore 8-bit (iso14443 starts with LSBit = bit 1)
+	// fixes wrong type for manufacturer Infineon (http://nfc-tools.org/index.php?title=ISO14443A)
+	byte sak = card->uid.sak & 0x7F;
```
the code in this function are a duplicate.
better:
```
PICC_Type type = PICC_GetType(sak);

// check for desfire
if( type == PICC_TYPE_ISO_14443_4 && card->atqa == 0x0344)
   type = PICC_TYPE_MIFARE_DESFIRE;
}

return type;

JPG-Consulting replied:
Yes, that's an option. :)

Rotzbua commented on MFRC522.cpp:
```diff
@@ -820,6 +820,62 @@ MFRC522::StatusCode MFRC522::PICC_HaltA() {
+ *
+ * @return STATUS_OK on success, STATUS_??? otherwise.
+ */
+MFRC522::StatusCode MFRC522::PICC_RATS(byte *bufferATS,	///< The buffer to store the ATS (Answer to select) in
```
equal to your desfire class: `PICC_RequestATS` ?

JPG-Consulting replied:
I think there are no changes apart from the name. In the DESFire library I choosed the long name, maybe more descriptive, and in the first versions I was using the abbreviature. 

Really the RATS is a bit more complex... In that method I'm using:

`bufferATS[1] = 0x50; // FSD=128, CID=0`

The byte sent represents the FSD and the CID. This is described here:
http://www.gorferay.com/data-transmission-protocol-iso-iec-14-433-4/

In the reply you would get additional information for the PPS command, which would allow to change the speed of the reader according to the parameters exchanged. I didn't want to go into tweaking the method too much as it may get a little confusing and not so sure if itt can bee nicelly representated (I'll look into it). Also, in my library I'm constantlly using CID 0 for the blocks, that is because CID = 0 has been hardcoded in the RATS method.

Maybe I could take a look into both those methods to get them to work with more flexibility... I was so focused on DESFire code I was happy with it doing both step to take me to the application layer DESFire. Recently I bumped into a wall :) I was testing all of this with a public transport card of my region and everything was working ok until the point I've hit encryption. I don't know the companies keys so I'm now waiting for some blank DESfire tags to arrive. I was also looking a little into 125 kHz cards... but it could be nice to find a way to finish off both of this methods completelly :) I'll look into them once more :)

JPG-Consulting replied:
Sorry... FSD=128 is a mistake... it should say FSD=64. It was hard coded that way to tell the card the maximum frame size our reader can handle, which is 64 if I'm not mistaken... The value however is correct 0x5X stands for FSD=64 bytes

JPG-Consulting commented:
Ok, I've changed the methods a little.

I've added a enum to set the baud rates for send and receive in the PPS command, also added an additional PPS command method to allow to send it with no params (Default params are kept). To keep thing tidy the PICC_ReadCardSerial will select 106Kbit/s as it is a decision each developper should make if other baud rates are available (Send in the ATS reply in byte TA1 if pressent).

Since the CID (Card ID) is hardcoded in the Request ATS method I choosed to remove that parameter from PPS to keep it a little more clear.

I've also added some extra comments to keep a little more documentation in the code.

JPG-Consulting commented:
Ok, I've added two basic methods:

TCL_Transceive
TCL_Deselect

This methods use ISO/IEC 14 443-3 I-Block and S-Block. The first method will support sending data to the PICC encapsulated in an I-Block (As expected by the standard). An overriden method should be created to add the optional NAD field. If anyone thinks of a better way to do it keeping the methods simple...

The Deselect metho does just that, Deselect a previously selected card (The card is selected on the ATS command). Currently the Card ID is hardcoded to "0", so there is not much use as it is used to read multiple card bound to a reader but i guess that would take plenty of additional changes as keeping track of bound cards... so i guess a hardcoded CID is just fine.

R-Blocks are currently not supported but i'll be looking to check the response in case it receives an R-Block to send the ACK (STATUS_OK) or NACK back.

JPG-Consulting commented:
I've made some changes to TCL methods.

I thought it would be more flexible if we could send a raw PCB Block... although it could be done by building the block and Transceiving data using one of the implemented methods such as PCD_TransceiveData I decided to create a PcbBlock structure containing all of the PCB block data (Except the epilogue, which is CRC). Inside the method I check if CRC generation is enabled in TX and RX registries and, if not, it is taken care of inside the method. It also simplifies a little the PCB Block generation as it will check if the PCB byte specifies the CID and/or NAD bytes are set and if they are they get added to the block (else they get ommitted).

By the way, the code in `PICC_PPS()` should have enabled the CRC generation in MFRC522. The additional check is just in case the developper has choosen to go another way.

In order to be able to send raw data I haven't implemented any additional checks in that method. For example, only I-Block should be able to send the NAD byte but thta's not checked here. 

There is another simpler method with different parameters which performs an I-Block transceive. Inside this method we perform some additional setup steps to make the method easier to use and just care about the application communication with an ISO 14 433-4 PICC ignoring all of the PCB Block process.

JPG-Consulting commented:
Right now the code seems to work ok... but I'm having some problems changing the speed between the PCD and the PICC. I'm able to send the PPS command ok, and everuthing seems to work out, however when I change to a higher speed I start to get timeouts and data is succesful just from time to time. Maybe there is an additional register that should be changed?

Any help would be appreciated as it would be the last minor step to the PPS command. Right now it works but it will stay at 106 kBits/s

Rotzbua commented:
@JPG-Consulting I merged your code into a testing branch. I will test this Frankenstein branch the next days. This pr was your latest version of the code right? Thanks for your great work!

Rotzbua commented:
@JPG-Consulting finally merged in master and released. I re-factored your implementations to a extension class. I hope it works for you. Best
